### ISSUE TITLE
Make lock-based cache stampede protection flexible and optional

### ISSUE THREAD

**Issue by pounard** [symfony/symfony#27730]
Following issue #27604 where @nicolas-grekas added the ``LockRegistry`` class to prevent cache stampede, and following #27729, I propose that the ``LockRegistry`` based cache stampede protection should be optionnaly disableable (without writing code) for various reasons:

- using an NFS shared filesystem, if it is configured aggressively for read velocity, may not always be in sync, making the lock mechanism behaviour potentially variable or inconsistent,
- using an NFS shared filesystem, which syncs correctly, instead of isolating each webhead from each other will in the opposite cause them to have concurrency problems with the locks,
- using a local filesystem for the code (using an immutable docker for example) but a shared file system for the generated cache (actual scenario we did put in place here on some infrastructures) we have an inconsistent locking mechanism: per default, ``LockRegistry`` will lock on the local isolated filesystem for generating shared files or shared caches, which makes it totally useless in the end,
- or just, using a slow filesystem, it will make useless I/O's that in most case, we're trying to get rid of (I take it that replacing it with a remote mutex, it will be slower, but sometime you just want to deactivate the protection).

Moreover, in my past, on many different sites typology we did had to implement mutexes for this exact same scenario: system cache generation (it's the case in Drupal, but we also did it in custom applications cases) but using a remote such as Redis or an SQL database to implement the mutexes. I give it to you, it's slower, but it's also much stronger as soon as you have multiple webheads, since both SQL and Redis are able to provide atomic complex operations, achieving a reliable mutex is quite easy.

The locking mechanism should get out of the ``GetTrait::get()`` method and be implemented as a proxy or decorator instead, allowing a much easier configurability or make it much easier to customize or replace stampede protection strategies.

**Comment by nicolas-grekas** [comment]
Thanks for raising, we need to have this discussion for sure!

> LockRegistry based cache stampede protection should be optionnaly disableable (without writing code)

to be sure I understand: you think we should have it enabled by default like now, but we should add some configuration system that could be expressed in config (yaml) to take control of the feature. Did I get it correctly?

Right now as you spotted, this is already a bit configurable but needs calling `LockRegistry::setFile()` somewhere (with an empty array to completely disable).

> NFS shared filesystem
> may not always be in sync, making the lock mechanism behaviour potentially variable or inconsistent

but stampede protection would still be reasonably safe - we're not contractually seeking for one and one warmup and having some low concurrency is still acceptable to me, isn't it?

> instead of isolating each webhead from each other will in the opposite cause them to have concurrency problems with the locks

which problems would this cause, apart from the slow down needed to acquire the lock?

> using a local filesystem for the code (using an immutable docker for example) but a shared file system for the generated cache (actual scenario we did put in place here on some infrastructures) we have an inconsistent locking mechanism: per default, LockRegistry will lock on the local isolated filesystem for generating shared files or shared caches, which makes it totally useless in the end,

I don't see the issue here, quite the contrary: as mentioned above, we're not providing any guarantee regarding concurrency. Instead, we're trying to have users wait for roughly one cache generation cycle while not killing the CPU with many parallel computations. Having this happen concurrently on different CPUs is fine to me.

> or just, using a slow filesystem, it will make useless I/O

ACK

> using a remote such as Redis or an SQL database to implement the mutexes. I give it to you, it's slower, but it's also much stronger as soon as you have multiple webheads, since both SQL and Redis are able to provide atomic complex operations, achieving a reliable mutex is quite easy.

ACK also: flock is per webhead and every webhead is computing the same cache item in parallel (but only once). From a user pov, and also from a CPU stampede pov, this is going to take the same time as a cluster-wide locking strategy. The difference pops in when cluster-wide CPU resources needs to be optimized, and also when having all webheads at work creates I/O stampede with the backend storages (needs quite a few webheads I suppose.) All of this can happen for sure.

> The locking mechanism should get out of the GetTrait::get() method and be implemented as a proxy or decorator instead

Could we build something based on the Lock component? Would you like to give it a try?

**Comment by pounard** [comment]
> to be sure I understand: you think we should have it enabled by default like now, but we should add some configuration system that could be expressed in config (yaml) to take control of the feature. Did I get it correctly?

And to go further, I think that an implementation based upon a decorator or a proxy would actually be simpler to understand, it seems like a responsibility soup that specific adapters know of the side stampede protection logic, they probably should not. Once that done instead of being hardcoded, (de)activation would be easy, either by manually altering the container using a custom services file, or using a compiler pass within the framework bundle itself. Same approach with probabilistic early expiration.

It would make the whole a lot easier to understand for developers, and make clearer of whose responsibility is each task: the backend/adapter should not care about all this, an external adapter developer should probably not even know about the GetTrait existance, it's internal.

**Comment by pounard** [comment]
> which problems would this cause, apart from the slow down needed to acquire the lock?

A slowdown :)

**Comment by pounard** [comment]
> I don't see the issue here, quite the contrary: as mentioned above, we're not providing any guarantee regarding concurrency. Instead, we're trying to have users wait for roughly one cache generation cycle while not killing the CPU with many parallel computations. Having this happen concurrently on different CPUs is fine to me.

Locking is then not guaranted anymore, multiple webheads can compute the same cache at the same time, and attempt to write the same files at the same moment, and over a NFS which has a slow sync, you can easily end-up, in the end with broken files or lost file operations.

**Comment by nicolas-grekas** [comment]
> attempt to write the same files at the same moment

wait, this reasoning involving files *written* to the fs is out of scope for the discussion: the lock never writes to files, it only locks. And FilesystemAdapter is safe from writing partial files because it uses "rename". So I'm not sure how your argumentation applies now.
